src/common/snippets/src/pass/extract_reshapes_from_mha.cpp

```diff
@@ -66,3 +67,93 @@ ov::snippets::pass::ExtractReshapesFromMHA::ExtractReshapesFromMHA() {
+        if (!static_shape_single_consumer(out)) {
+            return false;
+        }
+        const auto& out_shape = out.get_shape();
+        return ((out_shape.size() > 0) && (out_shape[0] == 1));
+    };
+    // reshape_2_m delete leading dimension of 1.
+    auto rank_reduction_reshape = [&](const ov::Output<ov::Node>& out) {
+        if (!static_shape_single_consumer(out)) {
+            return false;
+        }
+        auto in_shape_partial = out.get_node_shared_ptr()->get_input_partial_shape(0);
+        if (in_shape_partial.is_dynamic()) {
+            return false;
+        }
+        auto in_shape = in_shape_partial.to_shape();
+        if (in_shape.size() < 1 || in_shape[0] != 1) {
+            return false;
+        }
+        in_shape.erase(in_shape.begin());
+        const auto& out_shape = out.get_shape();
+        return out_shape == in_shape;
+    };
+
+    auto matmul_m = pattern::wrap_type<opset1::MatMul>(static_shape_single_consumer);
+    auto input_1_m = pattern::any_input(pattern::has_static_shape());
+    auto eltwise_1_m = pattern::optional<ov::op::util::BinaryElementwiseArithmetic>({matmul_m, input_1_m},
+                                                                                    static_shape_single_consumer);
+    auto reshape_1_m = pattern::wrap_type<opset1::Reshape>({eltwise_1_m, pattern::wrap_type<opset1::Constant>()},
+                                                           rank_upgrade_reshape);
+    auto input_2_m = pattern::any_input(has_leading_dimension_one);
+    auto eltwise_2_m = pattern::wrap_type<ov::op::util::BinaryElementwiseArithmetic>({reshape_1_m, input_2_m},
+                                                                                     static_shape_single_consumer);
+    auto reshape_2_m = pattern::wrap_type<opset1::Reshape>({eltwise_2_m, pattern::wrap_type<opset1::Constant>()},
+                                                           rank_reduction_reshape);
+
+    matcher_pass_callback callback = [OV_CAPTURE_CPY_AND_THIS](pattern::Matcher& m) {
+        OV_ITT_SCOPED_TASK(ov::pass::itt::domains::SnippetsTransform, "Snippets::op::RankUpgradeToRankReduction")
+        const auto& pattern_map = m.get_pattern_value_map();
+        const auto& matmul = pattern_map.at(matmul_m);
+        const auto matmul_node = ov::as_type_ptr<opset1::MatMul>(matmul.get_node_shared_ptr());
+        if (!ov::snippets::pass::TokenizeMHASnippets::is_matmul0_supported(matmul_node) ||
+            transformation_callback(matmul_node))
+            return false;
+
+        const auto& input_2 = pattern_map.at(input_2_m);
+        auto input_2_shape = input_2.get_shape();
+        input_2_shape.erase(input_2_shape.begin());
+        const auto target_shape = ov::opset1::Constant::create(ov::element::i32, {input_2_shape.size()}, input_2_shape);
+        const auto reshaped_input2 = std::make_shared<ov::opset1::Reshape>(input_2, target_shape, true);
```

Comment:
As far as I understand, if the second input of the second `Eltwise` op has shape `[1]`, this pass will match on this subgraph. And we will try to insert `Reshape` on the second input with empty target `input_2_shape`.

Could you please add check that inputs of the second `Eltwise` op have the same shape ranks? I think this check might be on `callback`.

Reply:
good catch, done, thanks Alexandra!

Reply:
Make sense to move the checks first and together, done, thanks Alexandra!